doc/guides/west/why.rst

```diff
@@ -6,15 +6,87 @@ History and Motivation
-  build system and debug mechanisms
-
-Additionally, it was desired that west be easily extensible by
-downstream users.
+* The ability to provide an extensible and user-friendly command-line interface
+  for basic Zephyr workflows
 
 During the development of west, a set of :ref:`west-design-constraints` were
 identified to avoid the common pitfalls of tools of this kind.
 
+Requirements
+************
+
+Although the motivation behind splitting the Zephyr codebase into multiple
+repositories is outside of the scope of this page, the fundamental requirements,
+along with a clear justification of the choice not to use existing tools and
+instead develop a new one, do belong here.
+
+The basic requirements are:
+
+* **R1**: Keep externally maintained code in separately maintained repositories
+  outside of the main zephyr repository, without requiring users to manually
+  clone each of the external repositories
+* **R2**: Provide a tool that both Zephyr users and distributors can make use of
+  to benefit from and extend
+* **R3**: Allow users and downstream distributions to override or remove
+  repositories without having to make changes to the zephyr repository
+* **R4**: Support both continuous tracking and commit-based (bisectable) project
+  updating
+
+Topologies supported
+********************
+
+The requirements above lead to three source code topologies supported by west:
+
+* **T1**: Star topology with zephyr as the manifest repository:
+
+  - The zephyr repository acts as the central repository and includes a
+    complete list of projects used upstream
+  - Default (upstream) configuration
+  - Analogy with existing mechanisms: Git submodules with zephyr as the
+    superproject
+
+* **T2**: Star topology with an application repository as the manifest
+  repository:
+
+  - A repository containing a Zephyr application acts as the central repository
+    and includes a complete list of other projects, including the zephyr
+    repository, required to build it
+  - Useful for those focused on a single application
+  - Analogy with existing mechanisms: Git submodules with the application as
+    the superproject, zephyr and other projects as submodules
+
+* **T3**: Forest topology:
+
+  - A dedicated manifest repository which contains no Zephyr source code,
+    and specifies a list of projects all at the same "level"
+  - Useful for downstream distributions with no "central" repository
+  - Analogy with existing mechanisms: Google repo-based source distribution
+
+Rationale for a custom tool
+***************************
+
+Existing tools were considered during west's initial design and development.
+None were found suitable for Zephyr's requirements. In particular, these were
+examined in detail:
+
+* Google repo
+
+  - Does not cleanly support using zephyr as the manifest repository (**R4**)
+  - Python 2 only
+  - Does not play well with Windows
+  - Assumes Gerrit is used for code review
+  - Poorly documented and maintained
```

Comment:
It's not professional to make disparaging comments like this, in official documentation, with zero fact or reference whatsoever to justify them.

FWIW I've used Google's repo for years and I wasn't in love with it but didn't find it "poorly documented and maintained". For sure it's being used on a production scale far beyond most other tools.

Reply:
`repo` for sure covers a quadrillion corner cases that we will never get to. It's very mature.

Reply:
> It's not professional to make disparaging comments like this, in official documentation, with zero fact or reference whatsoever to justify them.

@marc-hb:  this section is just being moved from existing documentation in repo-tool.rst; I didn't write it. I'll remove it because, on the whole, I agree it's not worth having in this document.

However, for the record, I agree with this statement 100%. Nothing against the repo developers, who are great, but the tool has been around for a long time and the quality of both documentation and maintenance is lacking in key areas.

Regarding maintenance: only 43 patches have been merged into master since January 1 2018, and that's not because there's nothing to do. Repo still doesn't support python 3 despite having had years to do so, and python 2 gets EOLed in 9 months.

I asked about this on the repo mailing list back in February 2018 and got nothing back:

https://groups.google.com/forum/#!topic/repo-discuss/MVOoA4WGYig

There's nothing obvious about when Python 3 is coming in the documentation etc. I am aware of `docs/python-support.md` in the repo sources; that's a plan without a timeline and no evidence of movement in the commit logs. This is certainly not what I would call active maintenance.

Regarding level of documentation: I've ramped up many new developers on repo for actual Android phone development, i.e. not some off-brand use case. And I have felt the pain of the lack of details in the documentation many times. The manifest format is extremely well documented, but the behavior of key commands like `repo sync` is very poorly documented. Here are the docs, for concreteness:

https://source.android.com/setup/develop/repo#sync

They say nothing about what the command does to the manifest repository (the answers are surprisingly complicated and nonintuitive), or what "syncing" does to the ref space inside each project (it's definitely *not* equivalent to `remote update` + `rebase` as claimed in the documentation; repo manages refs in each project somewhat similarly to how west uses `manifest-rev`). Similarly, documentation for `repo start` omits critical information about tracking branches. Etc.

> FWIW I've used Google's repo for years and I wasn't in love with it but didn't find it "poorly documented and maintained".

It's definitely worth a lot. I've also used repo for years, and have had the opposite experience, again FWIW.

> For sure it's being used on a production scale far beyond most other tools.

This is beyond dispute, but I must say has nothing to do with the quality of either its documentation or its level of maintenance.

Reply:
If you can summarize some of the points you just made then by all means add them to the list in the document.

Reply:
It's fine, as I said I agree with you that it's not worth having it in there on the whole. "Python 2 only" is already a deal-breaker.